test-utils/transaction-factory/src/complex_mode.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+/// This module implements the `MasterToNToM` mode:
+///
+/// Manufacture `num` transactions from the master account to `num`
+/// randomly created accounts. From each of these randomly created
+/// accounts manufacture a transaction to another randomly created
+/// account.
+/// Repeat this round `rounds` times. If `rounds` = 1 the behavior
+/// is the same as `MasterToN`.
+///
+///   A -> B
+///   A -> C
+///   A -> D
+///   ... x `num`
+///
+///   B -> E
+///   C -> F
+///   D -> G
+///   ...
+///   E -> H
+///   F -> I
+///   G -> J
+///   ...
+///   ... x `rounds`
+
+use std::sync::Arc;
+
+use log::info;
+use client::block_builder::api::BlockBuilder;
+use client::runtime_api::ConstructRuntimeApi;
+use sr_primitives::generic::BlockId;
+use sr_primitives::traits::{Block as BlockT, ProvideRuntimeApi, One, Zero};
+use substrate_service::{
+	FactoryBlock, FullClient, ServiceFactory, ComponentClient, FullComponents
+};
+
+use crate::{RuntimeAdapter, create_block};
+
+pub fn next<F, RA>(
+	factory_state: &mut RA,
+	client: &Arc<ComponentClient<FullComponents<F>>>,
+	prior_block_hash: <RA::Block as BlockT>::Hash,
+	prior_block_id: BlockId<F::Block>,
+) -> Option<<F as ServiceFactory>::Block>
+where
+	F: ServiceFactory,
+	F::RuntimeApi: ConstructRuntimeApi<FactoryBlock<F>, FullClient<F>>,
+	FullClient<F>: ProvideRuntimeApi,
+	<FullClient<F> as ProvideRuntimeApi>::Api: BlockBuilder<FactoryBlock<F>>,
+	RA: RuntimeAdapter,
+{
+	let total = factory_state.start_number() + factory_state.num() * factory_state.rounds();
+
+	if !(factory_state.block_no() < total && factory_state.round() < factory_state.rounds()) {
```

Comment:
```suggestion
	if factory_state.block_no() >= total || factory_state.round() >= factory_state.rounds() {
```

Reply:
I had also thought about this, but decided against it because both sides of the `||` need to be evaluated, whereas for my version the right side only is evaluated if the first is true.

Reply:
No? If the left is true, the right will not be evaluated for an or.

Reply:
Yours is also true, but this is mostly about readability.